### PR TITLE
android: correctly grab DNS settings

### PR DESCRIPTION
**Pull request by oxtoacart**
Closes #ENG-3005

I manually tested that this works on both Wifi and Cell, and that I'm able to switch Wifi off and on and maintain network connectivity.

### CODE REVIEW

File: android/src/main/java/com/tailscale/ipn/App.kt [hunk 1]
```diff
@@ -64,7 +64,6 @@ import java.util.Objects
 
 class App : Application(), libtailscale.AppContext {
   private val applicationScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
-  var dnsConfigObj = DnsConfig()
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/App.kt:67]: We had two instances of DnsConfig.

File: android/src/main/java/com/tailscale/ipn/App.kt [hunk 2]
```diff
@@ -75,11 +74,13 @@ class App : Application(), libtailscale.AppContext {
     private const val FILE_CHANNEL_ID = "tailscale-files"
     private const val FILE_NOTIFICATION_ID = 3
     private const val TAG = "App"
-    private val mainHandler = Handler(Looper.getMainLooper())
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/App.kt:78]: Unused

File: android/src/main/java/com/tailscale/ipn/App.kt [hunk 3]
```diff
@@ -75,11 +74,13 @@ class App : Application(), libtailscale.AppContext {
...
+    private val networkConnectivityRequest =
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/App.kt:77]: Moved from DnsConfig.

File: android/src/main/java/com/tailscale/ipn/App.kt [hunk 4]
```diff
@@ -143,33 +144,36 @@ class App : Application(), libtailscale.AppContext {
...
             super.onAvailable(network)
+
             val sb = StringBuilder()
             val linkProperties: LinkProperties? = connectivityManager.getLinkProperties(network)
-            val dnsList: MutableList<InetAddress> =
-                linkProperties?.getDnsServers() ?: mutableListOf()
+            val dnsList: MutableList<InetAddress> = linkProperties?.dnsServers ?: mutableListOf()
             for (ip in dnsList) {
-              sb.append(ip.getHostAddress()).append(" ")
+              sb.append(ip.hostAddress).append(" ")
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/App.kt:159]: Linter

File: android/src/main/java/com/tailscale/ipn/App.kt [hunk 5]
```diff
@@ -143,33 +144,36 @@ class App : Application(), libtailscale.AppContext {
...
+        networkConnectivityRequest,
         object : ConnectivityManager.NetworkCallback() {
           override fun onAvailable(network: Network) {
             super.onAvailable(network)
+
             val sb = StringBuilder()
             val linkProperties: LinkProperties? = connectivityManager.getLinkProperties(network)
-            val dnsList: MutableList<InetAddress> =
-                linkProperties?.getDnsServers() ?: mutableListOf()
+            val dnsList: MutableList<InetAddress> = linkProperties?.dnsServers ?: mutableListOf()
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/App.kt:157]: Linter

File: android/src/main/java/com/tailscale/ipn/App.kt [hunk 6]
```diff
@@ -143,33 +144,36 @@ class App : Application(), libtailscale.AppContext {
...
             for (ip in dnsList) {
-              sb.append(ip.getHostAddress()).append(" ")
+              sb.append(ip.hostAddress).append(" ")
             }
-            val searchDomains: String? = linkProperties?.getDomains()
+            val searchDomains: String? = linkProperties?.domains
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/App.kt:161]: Linter

File: android/src/main/java/com/tailscale/ipn/App.kt [hunk 7]
```diff
@@ -143,33 +144,36 @@ class App : Application(), libtailscale.AppContext {
...
             if (searchDomains != null) {
               sb.append("\n")
               sb.append(searchDomains)
             }
-            dnsConfigObj.updateDNSFromNetwork(sb.toString())
-            Libtailscale.onDnsConfigChanged()
+
+            if (dns.updateDNSFromNetwork(sb.toString())) {
```
Comment (oxtoacart) [inline android/src/main/java/com/tailscale/ipn/App.kt:167]: Avoid unnecessary changes, since calling onDnsConfigChanged() triggers a network reconfiguration on the backend.

File: android/src/main/java/com/tailscale/ipn/DnsConfig.java [hunk 8]
```diff
@@ -44,14 +41,14 @@ private String getDnsConfigs() {
         }
     }
 
-    void updateDNSFromNetwork(String dnsConfigs) {
+    boolean updateDNSFromNetwork(String dnsConfigs) {
         synchronized (this) {
-            this.dnsConfigs = dnsConfigs;
+            if (!dnsConfigs.equals(this.dnsConfigs)) {
```
Comment (ignoramous) [inline android/src/main/java/com/tailscale/ipn/DnsConfig.java:46]: for empty `dnsConfigs`, "localhost" on Android works; which usually forwards queries to either network-provided DNS (dhcp et al) or one of the pre-configured fallback DNSes as set by the OEM (usually `8.8.8.8`) on behalf of the VPN app.

Comment (oxtoacart) [reply android/src/main/java/com/tailscale/ipn/DnsConfig.java:46]: Thanks, I'll give that a try!

Comment (oxtoacart) [reply android/src/main/java/com/tailscale/ipn/DnsConfig.java:46]: Unfortunately, that did not work, at least not on a Samsung Galaxy S23.

**Review by agottardo — APPROVED** [review]
LGTM!

File: android/src/main/java/com/tailscale/ipn/DnsConfig.java (hunk 8 above)
Comment (ignoramous) [reply android/src/main/java/com/tailscale/ipn/DnsConfig.java:46]: Surprising. In our case, the VPN engine's (here, `Libtailscale`) stub resolver must forward incoming DNS packets it gets to `localhost:53` when it doesn't know the identity of underlying DNS.
